L1Trigger/L1TTwinMux/src/L1TTwinMuxAlgorithm.cc

```diff
@@ -0,0 +1,152 @@
+#include "L1Trigger/L1TTwinMux/interface/DTLowQMatching.h"
+#include "L1Trigger/L1TTwinMux/interface/RPCHitCleaner.h"
+#include "L1Trigger/L1TTwinMux/interface/IOPrinter.h"
+#include  "L1Trigger/L1TTwinMux/interface/L1MuTMChambPhContainer.h"
+
+using namespace std;
+
+
+void L1TTwinMuxAlgorithm::run(edm::Handle<L1MuDTChambPhContainer> inphiDigis,
+edm::Handle<L1MuDTChambThContainer> thetaDigis, edm::Handle<RPCDigiCollection> rpcDigis,const edm::EventSetup& c) {
+
+
+  ///ES Parameters
+  const L1TTwinMuxParamsRcd& tmParamsRcd = c.get<L1TTwinMuxParamsRcd>();
+  tmParamsRcd.get(tmParamsHandle);
+  const L1TTwinMuxParams& tmParams = *tmParamsHandle.product();
+  bool onlyRPC = tmParams.get_UseOnlyRPC();
+  bool onlyDT = tmParams.get_UseOnlyDT();
+  bool useLowQDT = tmParams.get_UseLowQDT();
+  bool correctBX = tmParams.get_CorrectDTBxwRPC();
+  bool verbose = tmParams.get_Verbose();
+
+
+
+  ///Align track segments that are coming in bx-1.
+  auto alignedDTs=std::make_shared<AlignTrackSegments>(*inphiDigis);
+  alignedDTs->run(c);
+  L1MuDTChambPhContainer phiDigis = alignedDTs->getDTContainer();
+  //if only DTs are required without bx correction
+  //return the aligned track segments
+  if(onlyDT && !correctBX && !useLowQDT) {
+    m_tm_phi_output = phiDigis;
+    return;
+  }
+  ///Clean RPC hits
+  auto rpcHitCl=std::make_shared<RPCHitCleaner>(*rpcDigis);
+  rpcHitCl->run(c);
+  RPCDigiCollection rpcDigisCleaned = rpcHitCl->getRPCCollection();
+
+  ///Translate RPC digis to DT primitives.
+  auto dt_from_rpc= std::make_shared<RPCtoDTTranslator>(rpcDigisCleaned);
+  dt_from_rpc->run(c);
+  L1MuDTChambPhContainer rpcPhiDigis = dt_from_rpc->getDTContainer();            //Primitves used for RPC->DT (only station 1 and 2)
+  L1MuDTChambPhContainer rpcHitsPhiDigis = dt_from_rpc->getDTRPCHitsContainer(); //Primitves used for bx correction
+
+  ///Match low q DT primitives with RPC hits in dphiWindow
+  auto dtlowq= std::make_shared<DTLowQMatching>(&phiDigis, rpcHitsPhiDigis);
+  dtlowq->run(c);
+
+  if(onlyDT && !correctBX && useLowQDT) {
+    m_tm_phi_output = phiDigis;
+    if(verbose){
+       IOPrinter ioPrinter;
+       cout<<"======DT========"<<endl;
+       ioPrinter.run(inphiDigis, m_tm_phi_output, rpcDigis, c);
+       cout<<"======RPC========"<<endl;
+       ioPrinter.run(&rpcHitsPhiDigis, m_tm_phi_output, &rpcDigisCleaned, c);
+       cout<<"+++++++++++++++++++++++++++++++++++++++++++++++++"<<endl;
+    }
+
+    return;
+  }
+
+
+  ///Correct(in bx) DT primitives by comparing them to RPC.
+//  DTRPCBxCorrection *rpc_dt_bx = new DTRPCBxCorrection(phiDigis,rpcHitsPhiDigis);
+  auto rpc_dt_bx=std::make_shared<DTRPCBxCorrection>(phiDigis,rpcHitsPhiDigis);
+  rpc_dt_bx->run(c);
+
+  L1MuDTChambPhContainer phiDigiscp = rpc_dt_bx->getDTContainer();
+
+  ///Add RPC primitives in case that there are no DT primitives.
+  std::vector<L1MuDTChambPhDigi> l1ttma_out;
+  L1MuDTChambPhDigi const* dtts1=nullptr;
```

Comment:
usual comments about variables and loops (why repeat this complex loop structure so many times?